### **UNSUPERVISED MACHINE LEARNING**
 Naturally in Machine Learning, We often implemement UnSupervised ML models as either; a part of the Feature Engineering Process,i.e to either help us obtain more insights during the EDA, or as a step to reduce the dimensions to improve the performance of any Supervised ML model. In this case scenario, we are implememnting it as part of the feature engineering process, in order to aid us obtaim nore insights from our data

For this class simulation/ Case study , we will use the pca for feature decomposition and kmeans as a clustering algorithm, please feel free to research and test implememnt 'other' decomposition algortihms listed above and other clustering algorithms explored above too, compare the performances and you can decide to stick with a combination that suits you


**ASSIGNMENT**

Using your Data Visualization and EDA Skills, Carry out Univariate, Bivariate and Multivariate Analysis EDA for this Data Set

In [1]:
#import required libraries
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb
import numpy as np

import warnings
warnings.filterwarnings("ignore")

### ASSESING DATA

In [2]:
# Load the data
data = pd.read_csv("..\Customer_Segmentation_\Acme.csv",encoding='UTF-8')
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [4]:
data.describe().T

count          mean          std       min       25%       50%  \
Quantity    541909.0      9.552250   218.081158 -80995.00      1.00      3.00   
UnitPrice   541909.0      4.611114    96.759853 -11062.06      1.25      2.08   
CustomerID  406829.0  15287.690570  1713.600303  12346.00  13953.00  15152.00   

                 75%      max  
Quantity       10.00  80995.0  
UnitPrice       4.13  38970.0  
CustomerID  16791.00  18287.0

In [5]:
data.describe(exclude=['int64','float64']).T

count unique                                 top    freq
InvoiceNo    541909  25900                              573585    1114
StockCode    541909   4070                              85123A    2313
Description  540455   4211  WHITE HANGING HEART T-LIGHT HOLDER    2369
InvoiceDate  541909  23260                    10/31/2011 14:41    1114
Country      541909     38                      United Kingdom  495478

In [6]:
#identifying cancelled orders
data.query("Quantity < 0")

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556     PLASTERS IN TIN CIRCUS PARADE       -12   
236      C536391     21984   PACK OF 12 PINK PAISLEY TISSUES       -24   
237      C536391     21983   PACK OF 12 BLUE PAISLEY TISSUES       -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

In [7]:
#identifying missing features
data.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [8]:
#identifying duplicated data points
data.duplicated().sum()

5268

#### observations
**After assessing our data set, we have outlined a couple of problems with the dataset;**
- wrong datatype, the invoice date should be a datetime object
- presence of missing data points, the customer id is an important feature, product description feature has some missing data points
- data duplicates, we have identified about 5268 data observations as duplicates
- we also have quite a numbe of cancelled orders, this sets up an interesting area that should be investigated

## Data normalization

Clustering algorithms (k-means clustering including) are highly affected by the scales of the data. So we need to [normalize](https://en.wikipedia.org/wiki/Normalization_(statistics) the data to be on the same scale.

##**Dimensionality Reduction**
 More input features often make a predictive modelling task more challenging to model, more generally referred to as the curse of dimensionality. thus,Dimensionality reduction refers to techniques that reduce the number of input variables in a dataset.

Principal Componenet Analysis (PCA) is a technique for redcuing the dimensions of a large dataset, increasing the interpreatblilty and at the same minimizing information loss
other exmaples of Dimensionality Reduction techniques include Self Organizing Maps (SOM), t-distributed Stochastic Neighbor Embedding (t-SNE) etc.


## **K-means clustering**

**K-means clustering** is a method of vector quantization, originally from signal processing, that aims to **partition n observations into k clusters in which each observation belongs to the cluster with the nearest mean** ([Wiki](https://en.wikipedia.org/wiki/K-means_clustering)). This is a method of **unsupervised learning** that learns the commonalities between groups of data without any target or labeled variable.

K-means clustering algorithm spits the records in the data into a **pre-defined number of clusters**, where the data points within each cluster are close to each other. One difficulty of using k-means clustering for customer segmentation is the fact that you need to know the number of clusters beforehand. Luckily, the silhouette coefficient can help you.

**The silhouette coefficient** measures how close the data points are to their clusters compared to other clusters. The silhouette coefficient values range from -1 to 1, where the closer the values are to 1, the better they are.

Let's find the best number of clusters:

### APPLYING CLUSTERING ALGO

### RECOMMENDATIONS AND CONCLUSION
